### PR TITLE
Refactored the code to include the Singelton class

### LINKED ISSUE #50: BustaMove instance is now static in GameScreen, does not solve anything
EDIT:
So this was due to the check of the YouWinScreen() is in BallManager, this should be moved to GameScreen.

Also the score is tracked in BallManager but written into an object present in the GameScreen, this is done directly because the BustaMove instance IN GameScreen is static. This should either be moved out of the BallManager or a reference to the ScoreKeeper should be passed to the BallManager just like the Cannon instance, as their relation is close, but you don't want the BallManager to rely upon the GameScreen class.

OLD:
It was already a reference, so why make this static? I get the idea of making this a variable which is shared across instances, but:
The constructor takes a reference to BustaMove, so this solves nothing, and only accomplishes half of what you want.

So we either change it back to just a normal private instance of BustaMove (my choice), or we create a second constructor for the GameScreen without the BustaMove instance, but then you get into the fact that you need to make sure you have an instance etc etc, and as we now only use one instance of GameScreen all this work is not worth it imo.

Also it's only in GameScreen and not in the other screens.

```
public static BustaMove game;
/*
 * ^^^^ made this object public static so it can be used in other classes 
 * (we would like to use this game instance because we're playing in it, without
 * creating a new one in the other class that's calling this object because that won't make any sense)
 * but, does making it "static" has any affect?
 */
```

### PR DESCRIPTION
**Pull request by 11maxed11**
Fixes #50

### CODE REVIEW

File: core/src/com/group66/game/screens/GameScreen.java [hunk 1]
```diff
@@ -52,13 +52,11 @@
...
      * 
-     * @param game
-     *            the game instance
      * @param randomLevel
      *            determines if a set level or a random level is used
      */
-    public GameScreen(BustaMove game, Boolean randomLevel) {
-        GameScreen.game = game;
+    public GameScreen(Boolean randomLevel) {
+        this.game = BustaMove.getGameInstance();
```
Comment (ALobker) [inline core/src/com/group66/game/screens/GameScreen.java]: We don't need a local variable if we have a singleton

File: core/src/com/group66/game/screens/HighScoreScreen.java [hunk 2]
```diff
@@ -28,10 +28,9 @@
     
     /**
      * Constructor for the high score screen
-     * @param game The game instance
      */
-    public HighScoreScreen(BustaMove game) {
-        this.game = game;
+    public HighScoreScreen() {
+        this.game = BustaMove.getGameInstance();
```
Comment (ALobker) [inline core/src/com/group66/game/screens/HighScoreScreen.java]: Same

File: core/src/com/group66/game/screens/MainMenuScreen.java [hunk 3]
```diff
@@ -32,12 +32,9 @@
...
     /**
      * Instantiates a new main menu screen.
-     *
-     * @param game
-     *            the game instance
      */
-    public MainMenuScreen(BustaMove game) {
-        this.game = game;
+    public MainMenuScreen() {
+        this.game = BustaMove.getGameInstance();
```
Comment (ALobker) [inline core/src/com/group66/game/screens/MainMenuScreen.java]: Same

File: core/src/com/group66/game/screens/SettingsScreen.java [hunk 4]
```diff
@@ -35,12 +35,9 @@
...
     /**
      * Instantiates a new main menu screen.
-     *
-     * @param game
-     *            the game instance
      */
-    public SettingsScreen(BustaMove game) {
-        this.game = game;
+    public SettingsScreen() {
+        this.game = BustaMove.getGameInstance();
```
Comment (ALobker) [inline core/src/com/group66/game/screens/SettingsScreen.java]: Same

File: core/src/com/group66/game/screens/SplitGameScreen.java [hunk 5]
```diff
@@ -53,14 +53,12 @@
...
-     * @param game
-     *            the game instance
+     *
      * @param randomLevel
      *            determines if a set level or a random level is used
      */
-    public SplitGameScreen(BustaMove game, Boolean randomLevel) {
-        SplitGameScreen.game = game;
+    public SplitGameScreen(Boolean randomLevel) {
+        this.game = BustaMove.getGameInstance();
```
Comment (ALobker) [inline core/src/com/group66/game/screens/SplitGameScreen.java]: Same

File: core/src/com/group66/game/screens/YouLoseScreen.java [hunk 6]
```diff
@@ -30,12 +30,9 @@
...
     /**
      * Instantiates a new main menu screen.
-     *
-     * @param game
-     *            the game instance
      */
-    public YouLoseScreen(BustaMove game) {
-        this.game = game;
+    public YouLoseScreen() {
+        this.game = BustaMove.getGameInstance();
```
Comment (ALobker) [inline core/src/com/group66/game/screens/YouLoseScreen.java]: Same

File: core/src/com/group66/game/screens/YouWinScreen.java [hunk 7]
```diff
@@ -30,12 +30,9 @@
...
     /**
      * Instantiates a new main menu screen.
-     *
-     * @param game
-     *            the game instance
      */
-    public YouWinScreen(BustaMove game) {
-        this.game = game;
+    public YouWinScreen() {
+        this.game = BustaMove.getGameInstance();
```
Comment (ALobker) [inline core/src/com/group66/game/screens/YouWinScreen.java]: Same

File: core/src/com/group66/game/BustaMove.java [hunk 8]
```diff
@@ -22,6 +22,21 @@
...
+    }
+    
+    /**
+     * Gets the only object available.
+     * 
+     * @return the only available object
+     */
+    public static BustaMove getGameInstance() {
+        return game;
+    }
```
Comment (jeroenoverman) [inline core/src/com/group66/game/BustaMove.java]: I think it's nicer to create the instance of the class here if it doesn't exist yet.  Just like in the slides of the lecture. (Design Patterns 2). Make it a synchronized method then, so you make sure there can't be more than 1 instance of it.

File: core/src/com/group66/game/screens/SplitGameScreen.java [hunk 9]
```diff
@@ -32,7 +32,7 @@
     }
     
     /** A place to store the game instance. */
-    public static BustaMove game;
+    private BustaMove game;
```
Comment (jeroenoverman) [inline core/src/com/group66/game/screens/SplitGameScreen.java]: You don't need this variable

File: core/src/com/group66/game/BustaMove.java (hunk 8 above)
Comment (11maxed11) [inline core/src/com/group66/game/BustaMove.java]: I made it just like in slide 11 in the DP2 lecture. This is another way to do it. Three ways can be used to describe the Singleton pattern, according to the lecture.

Comment (jeroenoverman) [inline core/src/com/group66/game/BustaMove.java]: Ok, but I thought the last method was the best method, but if they all work fine, this one is also fine to me.

**Comment by 11maxed11** [comment]
Okay guys, regarding the change of local variables:
If you take a look at the GameScreen class, we actually used the game instance many many times. For that, I thought it'd be better to put the instance in a local variable and then use that variable; this way this instance will be used many time with getting it only once.
For the remaining Screen classes, they actually need to use the instance once, but I followed the smae way to be consistent.

**Comment by jeroenoverman** [comment]
I don't think it will harm the program to use the getInstance method many many times, thats what its made for ;) especially if you implement the getInstance method the way you did. And by removing this local variable,  you use the singleton pattern the way it's designed to be used.

**Comment by 11maxed11** [comment]
Okayz then, I don't have a problem with that. Changes coming ahead.

**Comment by 11maxed11** [comment]
How about now?

**Comment by ALobker** [comment]
Works

**Pull request merged**